File: backend/app/models/twin.py
from __future__ import annotations

from datetime import datetime, UTC
from typing import Literal

from pydantic import BaseModel, Field


class TwinSignal(BaseModel):
    """A passive study signal from the extension or internal events."""

    source: Literal["chat", "ocr", "syllabus", "extension"] = "chat"
    concept_id: str | None = None
    topic_id: str | None = None
    url_hash: str | None = None
    domain: str | None = None
    similarity: float | None = None
    engagement: dict | None = None
    occurred_at: str = Field(default_factory=lambda: datetime.now(UTC).isoformat())


class VelocityPoint(BaseModel):
    """Single data point for learning velocity time-series."""

    week: str  # ISO week e.g. "2026-W33"
    concepts_learned: int = 0


class AcademicTwin(BaseModel):
    """Persistent model of a user's academic knowledge state."""

    user_id: str
    known_concepts: list[str] = Field(default_factory=list)
    weak_concepts: list[str] = Field(default_factory=list)
    missing_concepts: list[str] = Field(default_factory=list)
    velocity: list[VelocityPoint] = Field(default_factory=list)
    signals_count: int = 0
    last_updated: str = Field(default_factory=lambda: datetime.now(UTC).isoformat())


class ReadinessQuery(BaseModel):
    """Request: Can I learn concept X?"""

    concept: str = Field(description="The concept or topic the user wants to learn")


class PrerequisiteStatus(BaseModel):
    concept_id: str
    name: str
    status: Literal["ready", "missing"] = "missing"


class ReadinessResult(BaseModel):
    """Response: prerequisite breakdown for a target concept."""

    target_concept: str
    ready: bool = False
    ready_prerequisites: list[PrerequisiteStatus] = Field(default_factory=list)
    missing_prerequisites: list[PrerequisiteStatus] = Field(default_factory=list)
    summary: str = ""

File: frontend/lib/api/twin.ts
import { api } from "./client";

export interface VelocityPoint {
  week: string;
  concepts_learned: number;
}

export interface AcademicTwin {
  user_id: string;
  known_concepts: string[];
  weak_concepts: string[];
  missing_concepts: string[];
  velocity: VelocityPoint[];
  signals_count: number;
  last_updated: string;
}

export interface ReadinessResult {
  target_concept: string;
  ready: boolean;
  ready_prerequisites: { concept_id: string; name: string; status: string }[];
  missing_prerequisites: { concept_id: string; name: string; status: string }[];
  summary: string;
}

export async function fetchTwin(): Promise<AcademicTwin> {
  return api<AcademicTwin>("/api/twin");
}

export async function checkReadiness(concept: string): Promise<ReadinessResult> {
  return api<ReadinessResult>("/api/twin/readiness", {
    method: "POST",
    body: JSON.stringify({ concept }),
  });
}

export async function ingestSignal(signal: {
  source: string;
  concept_id?: string;
  topic_id?: string;
}): Promise<AcademicTwin> {
  return api<AcademicTwin>("/api/twin/signals", {
    method: "POST",
    body: JSON.stringify(signal),
  });
}
